### ISSUE TITLE
salt-ssh assumes root user if none specified in roster

### ISSUE THREAD

**Issue by tinder-nsoracco** [saltstack/salt#41256]
### Description of Issue/Question

`salt-ssh`, upon reading a roster file, and finding no `user:` key, will default to the user`root`, regardless of what might be present in `~/.ssh/config`.  I am running `salt-ssh` as a non-privileged user, and expected it to work with what was in my config for the host.  Instead, it guesses I mean root if it finds no `user:` entry.

### Setup

`salt-ssh` runs as a regular user on an admin's workstation.  The remote target is a basic Ubuntu VM.  I have configured a user for me to use, and configured my local `~/.ssh/config` to permit me to log in via a simple `ssh salttest` invocation.  The config contains information that might otherwise be specified in a roster file, were it necessary to do so.

### Steps to Reproduce Issue

1. Create a copy of `/etc/salt` and modify to handle all of your paths, etc., to ensure you have an otherwise correct salt-ssh installation.

2. Create a roster file as follows:

```
salttest:
  host: salttest
  sudo: True
  priv: agent-forwarding
```

3. Configure `~/.ssh/config` as follows:
```
host salttest:
  user: my.user
  hostname: salttest.example.com
  port: 2222
```

4. Create a basic Saltfile in your copy of `/etc/salt` as follows
```
salt-ssh:
  config_dir: ~/path/to/etc/salt
  cachedir: /var/tmp/salt
  log_file: /var/tmp/salt/sshlog
```
5. Load any crypto into `ssh-agent`

6. verify that `ssh salttest` works as expected.

7. try `salt-ssh -c /path/to/Saltfile salttest test.ping`

8. Assuming it fails, add the correct username into the roster file, and try again.  It should succeed.

#### Expected outcome
```
salttest:
    True
```
#### Command issued by salt-ssh (when configured with a user: key)
```
execve("/usr/bin/ssh", ["ssh", "salttest", "-o", "KbdInteractiveAuthentication=no", "-o", "PasswordAuthentication=no", "-o", "GSSAPIAuthentication=no", "-o", "ConnectTimeout=65", "-o", "Port=2222", "-o", "IdentityFile=agent-forwarding", "-o", "User=my.user", "/bin/sh"], [/* 69 vars */]) = 0
```

#### Actual outcome
```
salttest:
    ----------
    retcode:
        255
    stderr:
        no such identity: agent-forwarding: No such file or directory
        Permission denied (publickey).
    stdout:
```
#### Command issued by salt-ssh (with no user configured)
```
execve("/usr/bin/ssh", ["ssh", "salttest", "-o", "KbdInteractiveAuthentication=no", "-o", "PasswordAuthentication=no", "-o", "GSSAPIAuthentication=no", "-o", "ConnectTimeout=65", "-o", "Port=2222", "-o", "IdentityFile=agent-forwarding", "-o", "User=root", "/bin/sh"], [/* 69 vars */]) = 0
```
### Versions Report

Salt Version:
           Salt: 2016.11.0

Dependency Versions:
           cffi: 1.7.0
       cherrypy: Not Installed
       dateutil: 2.6.0
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.3
          ioflo: Not Installed
         Jinja2: 2.8.1
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, May 10 2017, 20:04:28)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.2
            ZMQ: 4.1.4

System Versions:
           dist: fedora 25 Twenty Five
        machine: x86_64
        release: 4.10.14-200.fc25.x86_64
         system: Linux
        version: Fedora 25 Twenty Five

**Comment by shallot** [comment]
I wouldn't necessarily go so far as to immediately expect salt-ssh to include local ssh config, but it stands to reason that it should absolutely not take `strace` to get salt-ssh to give out the information about what implicit setting it used after the most basic operation breaks.

Maybe the problem is that people do what the manual tells them - `test.ping` yet there could probably be a more user-friendly command to use for bootstrapping salt-ssh, one that would make more of an effort to explain to the user what's going on.

**Comment by tinder-nsoracco** [comment]
I agree, it would be nice to see what's happening without needing `strace`. ;)  Perhaps there's a lower level than `-ldebug` I could've used that would've obviated the need for strace.  I'll dig a little deeper today.

I also agree that `salt-ssh` shouldn't bother to read in and parse the local ssh_config, that's `ssh`'s job after all.

Personally, I feel that if a parameter is missing, I think the default behavior should be to assume that the user knows what they're doing with their ssh config and not try and choose "sane" defaults.  Or, maybe a sentinel value like: `user: sshconfig` that tells `salt-ssh` on a per-parameter basis to disengage any sane default logic?  (Or maybe that needs to be at the whole roster level, I really don't know.)

**Comment by Ch3LL** [comment]
@tinder-nsoracco does this feature request grasp what you are trying to accomplish: https://github.com/saltstack/salt/issues/35727

Also if you use `-ltrace` with salt-ssh you can usually see all the ssh calls its making. Thats how i debug salt-ssh atleast.

**Comment by tinder-nsoracco** [comment]
@Ch3LL : thanks for the {{-ltrace}} flag, I'll try that shortly. 👍 
As to #35727, assuming that any gaps discovered in reading in the config file don't result in {{salt-ssh}} making decisions for me, then yes, probably. :)  

Of course handling that file when you have 1000s of hosts gets interesting.  For example, how would 'salt-ssh' handle parsing `host *` entries, or other wildcard `host` patterns?  I would guess that the shortest path to the desired outcome, for me at least, would be to remove the code that assumes "root" for "user" when its absent.  Reading `ssh_config` is probably more work once you take wildcards into account.

**Comment by tinder-nsoracco** [comment]
Then again, I don't use `salt-ssh` the way most people would use ansible or some other "agent-less" CM system.  I just have a couple systems that can't connect to my salt-master that I still have to manage.  `salt-ssh` is fantastic for that, so thank you.  I would imagine that other users who use it more widely than I do might have more to say about this issue.

**Comment by shallot** [comment]
BTW, there was recently a related question at http://stackoverflow.com/questions/43565258/how-to-configure-salt-sshs-roster - even if salt-ssh doesn't implicitly support the whole `ssh_config` functionality, it could have a few common configurable settings, notably the user.
